# Automating data analysis

In the previous chapters, you learned how to perform each step of the metagenomic data analysis pipeline manually. While this is a valuable learning experience, it's not practical for analyzing large datasets or for ensuring reproducibility in the long term.

We can make use of a tool called [Snakemake](https://snakemake.readthedocs.io/en/stable/) to automate the previous steps into a single pipeline. With Snakemake, you can define the steps of your analysis in a `Snakefile` and then let Snakemake handle the execution, dependency management, and error handling.

## Preparing to run the workflow.

To run the automated workflow, you'll need to make sure that your project directory is set up correctly and that you have the necessary Conda environment activated.

In the command line, navigate to your project directory. First check your current directory with the pwd command:

``` bash
pwd
```

Change your current directory using cd if needed:

``` bash
cd /{folder1}/{folder2} # Replace with the actual path to your project directory
```

Next, run the `ls` command to list the files in the current directory and check if the `illumina_snakemake.smk` file is present. This is the "recipe" for the workflow, describing all the steps we have done by hand, and it is most commonly placed in the root directory of your project (you can open the `.smk` file with a text editor and have a look).

Make sure that we have activated the appropriate conda environment and check if the command line has `(stable_env_1)` in front.

``` bash
conda activate stable_env_1 # Replace environment name is needed
```

Lastly, check the `raw_data` directory. Make sure that you have a `raw_data/` directory in your project folder and that it contains your gzipped FASTQ files (e.g., `sample1_R1_001.fastq.gz`, `sample1_R2_001.fastq.gz`, etc).

## Running the workflow

After setting everything up, we can redo the analysis for all samples in a single step. First we will test out a dry run to see if any errors appear. A dry run will not execute any of the commands but will instead display what would be done. This will help identify any errors in the Snakemake or config file.

``` bash
snakemake \
--snakefile \
illumina_snakemake.smk \
--use-conda \
--cores {threads} \
--dryrun
```

If no errors appear, then remove the `--dryrun` argument and run it again to fully execute the workflow.
